tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -297,6 +298,239 @@ StatusOr<bool> FuseConvAlpha(HloComputation* comp) {
+  }
+}
+
+// Captures the graph of pointwise operations operating on the convolution.
+bool CaptureConvGraph(HloInstruction* instr,
+                      std::vector<HloInstruction*>& operands,
+                      GraphString& graph_string, HloInstruction*& final_instr,
+                      HloInstruction* x_scale, HloInstruction* w_scale,
+                      bool x_mult_scale, bool w_mult_scale) {
+  absl::flat_hash_set<int> visited_instrs;
+  graph_string.AppendOp("conv", instr->shape().element_type());
+
+  // Shift the scaling of the inputs to the output of the convolution.
+  if (x_scale && w_scale && x_mult_scale == w_mult_scale) {
+    HloInstruction* product =
+        instr->AddInstruction(HloInstruction::CreateBinary(
+            x_scale->shape(), HloOpcode::kMultiply, x_scale, w_scale));
+    operands.push_back(product);
+    graph_string.AppendOp(x_mult_scale ? "scale" : "invscale",
+                          instr->shape().element_type());
+  } else {
+    if (x_scale) {
+      operands.push_back(x_scale);
+      graph_string.AppendOp(x_mult_scale ? "scale" : "invscale",
+                            instr->shape().element_type());
+    }
+    if (w_scale) {
+      operands.push_back(w_scale);
+      graph_string.AppendOp(w_mult_scale ? "scale" : "invscale",
+                            instr->shape().element_type());
+    }
+  }
+  return CaptureConvGraphRecursive(instr, operands, graph_string,
+                                   visited_instrs, final_instr);
+}
+
+// Matches convolutions operating on FP8 inputs and filters and rewrites into a
+// ForwardGraph Custom Call. For scaled FP8 convolutions on Hopper systems, the
+// following steps are elided and rewritten into a ForwardGraph Custom Call:
+//
+// 1. Cast the filter and input from FP8 to a wider type such as FP16 or FP32.
+// 2. Unscale the filter and input by multiplying them by the corresponding
+// input scale.
+// 3. Evaluate the convolution based on the scaled filter and input.
+// 4. Optionally add a matrix bias to the result and apply a ReLU activation.
+// 5. Scale the output by dividing the output by the output scale.
+// 6. Cast the output back to FP8.
+
+StatusOr<bool> F8GraphConv(HloComputation* comp, se::CudaComputeCapability cc) {
+  bool changed = false;
+#if CUDA_VERSION >= 12000
+  for (auto instr : comp->MakeInstructionPostOrder()) {
+    const DebugOptions& debug_options =
+        instr->GetModule()->config().debug_options();
+    if (!cc.IsAtLeast(se::CudaComputeCapability::HOPPER)) {
+      return false;
+    }
+    HloInstruction *convolution, *gte, *input, *filter, *final_instr,
+        *x_scale = nullptr, *w_scale = nullptr, *x_scale_op = nullptr,
+        *w_scale_op = nullptr;
+    std::vector<HloInstruction*> operands;
+    GraphString graph_string;
+
+    // TODO(philipphack): Consider allowing ops between dequantization and
+    // convolution.
+    auto pattern = m::GetTupleElement(
+        &gte,
+        m::CustomCall(
```

Comment:
Does cuDNN allow any combination of FP8 types? If not, check for the correct combination.

Also add a test for two E5M2 inputs (which cuBLAS LT doesn't support) and one E5M2, one E4M3 input.